internal/hcl/graph.go

```diff
@@ -0,0 +1,325 @@
+			if strings.HasPrefix(ref.Key, "data.") && len(parts) >= 3 {
+				// Source ID is the first 3 parts or less if the length of parts is less than 3
+				idx = 3
+			} else if len(parts) >= 2 {
+				// resources and outputs don't have a type prefix in the references
+				idx = 2
+			}
+			srcId = strings.Join(parts[:idx], ".")
+
+			// Don't add the module prefix for providers since they are
+			// evaluated in the root module
+			if !strings.HasPrefix(srcId, "provider.") && ref.ModuleAddress != "" {
+				srcId = fmt.Sprintf("%s.%s", ref.ModuleAddress, srcId)
+			}
+
+			// If the reference points to a different module then we want to add
+			// a dependency for that module call
+			if ref.ModuleAddress != vertex.ModuleAddress() {
+				srcId = ref.ModuleAddress
+			}
+
+			// Strip the count/index suffix from the source ID
+			srcId = stripCount(srcId)
+
+			if srcId == vertex.ID() {
+				continue
+			}
+
+			g.logger.Debugf("adding edge: %s, %s", srcId, vertex.ID())
+			err := g.dag.AddEdge(srcId, vertex.ID())
+			if err != nil {
+				g.logger.Debugf("error adding edge: %s", err)
+			}
+		}
+	}
+
+	// Setup initial context
+	evaluator.ctx.Set(cty.ObjectVal(map[string]cty.Value{}), "var")
+	evaluator.ctx.Set(cty.ObjectVal(map[string]cty.Value{}), "data")
+	evaluator.ctx.Set(cty.ObjectVal(map[string]cty.Value{}), "local")
+	evaluator.ctx.Set(cty.ObjectVal(map[string]cty.Value{}), "output")
+
+	// TODO: is there a better way of doing this?
+	// Add the locals block to the list of filtered blocks. These won't get
+	// added when we walk the graph since locals are evaluated per attribute
+	// and not per block, so we need to make sure this is done here.
+	evaluator.AddFilteredBlocks(evaluator.module.Blocks.OfType("locals")...)
+
+	return nil
+}
+
+func (g *Graph) AsJSON() ([]byte, error) {
+	return g.dag.MarshalJSON()
+}
+
+func (g *Graph) Walk() {
+	v := NewGraphVisitor(g.logger)
+
+	flowCallback := func(d *dag.DAG, id string, parentResults []dag.FlowResult) (interface{}, error) {
+		vertex, _ := d.GetVertex(id)
+
+		v.Visit(id, vertex)
+
+		return vertex, nil
+	}
+
+	_, _ = g.dag.DescendantsFlow(g.rootVertex.ID(), nil, flowCallback)
+}
+
+type GraphVisitor struct {
+	logger *logrus.Entry
+}
+
+func NewGraphVisitor(logger *logrus.Entry) *GraphVisitor {
+	return &GraphVisitor{
+		logger: logger,
+	}
+}
+
+func (v *GraphVisitor) Visit(id string, vertex interface{}) {
+	v.logger.Debugf("visiting %s", id)
+
+	vert := vertex.(Vertex)
+
+	visitMu.Lock()
+	err := vert.Evaluate()
+	visitMu.Unlock()
+	if err != nil {
+		v.logger.WithError(err).Debugf("could not evaluate %s ignoring", id)
+		return
+	}
+
+	blocks, err := vert.Expand()
+	if err != nil {
+		v.logger.WithError(err).Debugf("could not expand %s ignoring", id)
+		return
+	}
+
+	ve := vert.Evaluator()
+	if ve != nil && vert.Evaluator() != nil {
+		vert.Evaluator().AddFilteredBlocks(blocks...)
+	}
+}
+
+func referencesForBlock(b *Block) []VertexReference {
```

Comment:
I'd prefer if we moved this and the corresponding `referencesForAttribute` to methods on the `Block` and `Attribute` types.